# Unit 1 Assignment: The Model Benchmark Challenge

**Objective**: In this assignment, you will step beyond simply using a model and instead evaluate the architectural differences between **BERT**, **RoBERTa**, and **BART**. You will force these models to perform tasks they might not be designed for, to observe why architecture matters.

---

## The Models to Test
1.  **BERT** (`bert-base-uncased`): An **Encoder-only** model (designed for understanding, not generation).
2.  **RoBERTa** (`roberta-base`): An optimized **Encoder-only** model.
3.  **BART** (`facebook/bart-base`): An **Encoder-Decoder** model (designed for seq2seq tasks like translation/generation).

## Setup: Import Libraries

First, we'll import the necessary libraries and set up device configuration for GPU acceleration.

In [1]:
from transformers import pipeline
import torch
import warnings
warnings.filterwarnings('ignore')

C:\Users\stard\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


### Check GPU Availability

Let's check if CUDA is available for GPU acceleration.

In [2]:
# Set up device for GPU acceleration
device = 0 if torch.cuda.is_available() else -1
if device == 0:
    print(f"✓ Using GPU: {torch.cuda.get_device_name(0)}")
else:
    print("✗ CUDA not available, using CPU")

✓ Using GPU: NVIDIA GeForce RTX 4060 Laptop GPU


---

## Experiment 1: Text Generation

**Task**: Try to generate text using the prompt: `"The future of Artificial Intelligence is"`

**Hypothesis**: Which models will fail? Why? (Hint: Can an Encoder *generate* new tokens easily?)

In [3]:
prompt = "The future of Artificial Intelligence is"

print("="*60)
print("EXPERIMENT 1: TEXT GENERATION")
print("="*60)

EXPERIMENT 1: TEXT GENERATION


### Test 1.1: BERT for Text Generation

In [15]:
print("\n--- Testing BERT (bert-base-uncased) ---")
try:
    bert_gen = pipeline('text-generation', model='bert-base-uncased', device=device)
    result_bert = bert_gen(prompt, max_length=30)
    print(f"Result: {result_bert[0]['generated_text']}")
    print("Status: SUCCESS")
except Exception as e:
    print(f"Status: FAILURE")
    print(f"Error: {type(e).__name__}: {str(e)[:100]}")


--- Testing BERT (bert-base-uncased) ---


If you want to use `BertLMHeadModel` as a standalone, add `is_decoder=True.`
Device set to use cuda:0
Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Both `max_new_tokens` (=256) and `max_length`(=30) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Result: The future of Artificial Intelligence is................................................................................................................................................................................................................................................................
Status: SUCCESS


### Test 1.2: RoBERTa for Text Generation

In [5]:
print("\n--- Testing RoBERTa (roberta-base) ---")
try:
    roberta_gen = pipeline('text-generation', model='roberta-base', device=device)
    result_roberta = roberta_gen(prompt, max_length=30)
    print(f"Result: {result_roberta[0]['generated_text']}")
    print("Status: SUCCESS")
except Exception as e:
    print(f"Status: FAILURE")
    print(f"Error: {type(e).__name__}: {str(e)[:100]}")


--- Testing RoBERTa (roberta-base) ---


If you want to use `RobertaLMHeadModel` as a standalone, add `is_decoder=True.`
Device set to use cuda:0
Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Both `max_new_tokens` (=256) and `max_length`(=30) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Result: The future of Artificial Intelligence is
Status: SUCCESS


### Test 1.3: BART for Text Generation

In [6]:
print("\n--- Testing BART (facebook/bart-base) ---")
try:
    bart_gen = pipeline('text-generation', model='facebook/bart-base', device=device)
    result_bart = bart_gen(prompt, max_length=30)
    print(f"Result: {result_bart[0]['generated_text']}")
    print("Status: SUCCESS")
except Exception as e:
    print(f"Status: FAILURE")
    print(f"Error: {type(e).__name__}: {str(e)[:100]}")


--- Testing BART (facebook/bart-base) ---


Some weights of BartForCausalLM were not initialized from the model checkpoint at facebook/bart-base and are newly initialized: ['lm_head.weight', 'model.decoder.embed_tokens.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Device set to use cuda:0
Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Both `max_new_tokens` (=256) and `max_length`(=30) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Result: The future of Artificial Intelligence is Carr Carr Carr asc creating Mapsiblstice guaranteed guaranteed person personrieved antagonistObj antagonistottest shortenHTottest delinquObjObj rugged PROC PROC PROC wiseKnow Eclipsebyssbyss SiegeutedObjottestObjRingRingRing shortenObjObj lickingObjiblRing ascEvery asc RicanoenixObj asc Economist asc uninsuredibliblibl ascObjottestibliblObj Rican asc spons guaranteediblibl Ud ascRing Economist wise wise ascbernatorialVan ascEveryEveryEveryaque ascibl contributingibl asciblbernatorialibl IzanEvery Sah gonnaBottomObjEveryEveryObj ascObj ascibl Sah Sahuated ascEvery overly Sah Sah HAMObjdraftEveryObjEvery asc LOW IzanEveryMEN veteransObj gonnaEveryibl BrotherhoodEveryEvery shortenibl Legislation asc ascursEvery IzanObjEveryiblObjizaEveryuatedrieriblEveryiblEvery asc pertinentibl HAMEveryEveryizadraftibl UdEvery maybeibluatedallsEveryurs assailantEveryEveryMEN Izaniblibl Economist IzanEvery Izan ascEvery guaranteed asc IzanObjObj HAMEveryzzy

---

## Experiment 2: Masked Language Modeling (Fill-in-the-Blank)

**Task**: Predict the missing word in: `"The goal of Generative AI is to [MASK] new content."`

**Hypothesis**: This is how BERT/RoBERTa were trained. They should perform well.

In [7]:
# Note: Different models use different mask tokens
masked_sentence_bert = "The goal of Generative AI is to [MASK] new content."
masked_sentence_roberta = "The goal of Generative AI is to <mask> new content."

print("="*60)
print("EXPERIMENT 2: MASKED LANGUAGE MODELING")
print("="*60)

EXPERIMENT 2: MASKED LANGUAGE MODELING


### Test 2.1: BERT for Fill-Mask

In [8]:
print("\n--- Testing BERT (bert-base-uncased) ---")
try:
    bert_mask = pipeline('fill-mask', model='bert-base-uncased', device=device)
    result_bert = bert_mask(masked_sentence_bert)
    print("Top 3 predictions:")
    for i, pred in enumerate(result_bert[:3], 1):
        print(f"  {i}. {pred['token_str']} (score: {pred['score']:.4f})")
    print("Status: SUCCESS")
except Exception as e:
    print(f"Status: FAILURE")
    print(f"Error: {type(e).__name__}: {str(e)[:100]}")


--- Testing BERT (bert-base-uncased) ---


Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Device set to use cuda:0


Top 3 predictions:
  1. create (score: 0.5397)
  2. generate (score: 0.1558)
  3. produce (score: 0.0541)
Status: SUCCESS


### Test 2.2: RoBERTa for Fill-Mask

In [9]:
print("\n--- Testing RoBERTa (roberta-base) ---")
try:
    roberta_mask = pipeline('fill-mask', model='roberta-base', device=device)
    result_roberta = roberta_mask(masked_sentence_roberta)
    print("Top 3 predictions:")
    for i, pred in enumerate(result_roberta[:3], 1):
        print(f"  {i}. {pred['token_str']} (score: {pred['score']:.4f})")
    print("Status: SUCCESS")
except Exception as e:
    print(f"Status: FAILURE")
    print(f"Error: {type(e).__name__}: {str(e)[:100]}")


--- Testing RoBERTa (roberta-base) ---


Device set to use cuda:0


Top 3 predictions:
  1.  generate (score: 0.3711)
  2.  create (score: 0.3677)
  3.  discover (score: 0.0835)
Status: SUCCESS


### Test 2.3: BART for Fill-Mask

In [10]:
print("\n--- Testing BART (facebook/bart-base) ---")
try:
    bart_mask = pipeline('fill-mask', model='facebook/bart-base', device=device)
    result_bart = bart_mask(masked_sentence_roberta)  # BART uses <mask> like RoBERTa
    print("Top 3 predictions:")
    for i, pred in enumerate(result_bart[:3], 1):
        print(f"  {i}. {pred['token_str']} (score: {pred['score']:.4f})")
    print("Status: SUCCESS")
except Exception as e:
    print(f"Status: FAILURE")
    print(f"Error: {type(e).__name__}: {str(e)[:100]}")


--- Testing BART (facebook/bart-base) ---


Device set to use cuda:0


Top 3 predictions:
  1.  create (score: 0.0746)
  2.  help (score: 0.0657)
  3.  provide (score: 0.0609)
Status: SUCCESS


---

## Experiment 3: Question Answering

**Task**: Answer the question `"What are the risks?"` based on the context: `"Generative AI poses significant risks such as hallucinations, bias, and deepfakes."`

**Note**: Using a "base" model (not fine-tuned for SQuAD) might yield random or poor results. Observe this behavior.

In [11]:
context = "Generative AI poses significant risks such as hallucinations, bias, and deepfakes."
question = "What are the risks?"

print("="*60)
print("EXPERIMENT 3: QUESTION ANSWERING")
print("="*60)
print(f"Context: {context}")
print(f"Question: {question}")

EXPERIMENT 3: QUESTION ANSWERING
Context: Generative AI poses significant risks such as hallucinations, bias, and deepfakes.
Question: What are the risks?


### Test 3.1: BERT for Question Answering

In [12]:
print("\n--- Testing BERT (bert-base-uncased) ---")
try:
    bert_qa = pipeline('question-answering', model='bert-base-uncased', device=device)
    result_bert = bert_qa(question=question, context=context)
    print(f"Answer: {result_bert['answer']}")
    print(f"Score: {result_bert['score']:.4f}")
    print("Status: SUCCESS")
except Exception as e:
    print(f"Status: FAILURE")
    print(f"Error: {type(e).__name__}: {str(e)[:100]}")


--- Testing BERT (bert-base-uncased) ---


Some weights of BertForQuestionAnswering were not initialized from the model checkpoint at bert-base-uncased and are newly initialized: ['qa_outputs.bias', 'qa_outputs.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Device set to use cuda:0


Answer: and deepfakes.
Score: 0.0050
Status: SUCCESS


### Test 3.2: RoBERTa for Question Answering

In [13]:
print("\n--- Testing RoBERTa (roberta-base) ---")
try:
    roberta_qa = pipeline('question-answering', model='roberta-base', device=device)
    result_roberta = roberta_qa(question=question, context=context)
    print(f"Answer: {result_roberta['answer']}")
    print(f"Score: {result_roberta['score']:.4f}")
    print("Status: SUCCESS")
except Exception as e:
    print(f"Status: FAILURE")
    print(f"Error: {type(e).__name__}: {str(e)[:100]}")


--- Testing RoBERTa (roberta-base) ---


Some weights of RobertaForQuestionAnswering were not initialized from the model checkpoint at roberta-base and are newly initialized: ['qa_outputs.bias', 'qa_outputs.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Device set to use cuda:0


Answer: significant risks such as hallucinations, bias, and deepfakes
Score: 0.0119
Status: SUCCESS


### Test 3.3: BART for Question Answering

In [14]:
print("\n--- Testing BART (facebook/bart-base) ---")
try:
    bart_qa = pipeline('question-answering', model='facebook/bart-base', device=device)
    result_bart = bart_qa(question=question, context=context)
    print(f"Answer: {result_bart['answer']}")
    print(f"Score: {result_bart['score']:.4f}")
    print("Status: SUCCESS")
except Exception as e:
    print(f"Status: FAILURE")
    print(f"Error: {type(e).__name__}: {str(e)[:100]}")


--- Testing BART (facebook/bart-base) ---


Some weights of BartForQuestionAnswering were not initialized from the model checkpoint at facebook/bart-base and are newly initialized: ['qa_outputs.bias', 'qa_outputs.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
Device set to use cuda:0


Answer: Generative AI poses significant risks such as hallucinations,
Score: 0.0371
Status: SUCCESS


---

## Deliverable: Observation Table

Based on your experiments above, fill out this table with your observations.

| Task | Model | Classification (Success/Failure) | Observation (What actually happened?) | Why did this happen? (Architectural Reason) |
| :--- | :--- | :--- | :--- | :--- |
| **Generation** | BERT | Success (Poor Quality) | Generated repetitive dots (...) instead of meaningful text | BERT is an Encoder-only model not designed for autoregressive generation; it was forced to work with `is_decoder=True` but lacks proper training for next-token prediction |
| | RoBERTa | Success (No Output) | Returned the input prompt unchanged with no generation | RoBERTa is an Encoder-only model optimized for understanding tasks, not text generation; lacks the decoder architecture needed for autoregressive generation |
| | BART | Success (Poor Quality) | Generated random nonsensical tokens (e.g., "Carr", "Izan", "antagonist") | BART-base is an Encoder-Decoder but the causal LM head was randomly initialized (not trained), resulting in meaningless output |
| **Fill-Mask** | BERT | Success | Correctly predicted "create" (53.97%), "generate" (15.58%), "produce" (5.41%) | BERT is explicitly trained on Masked Language Modeling (MLM); this is its core training objective |
| | RoBERTa | Success | Correctly predicted "generate" (37.11%), "create" (36.77%), "discover" (8.35%) | RoBERTa is an optimized BERT variant also trained on MLM; performs well on this task |
| | BART | Success (Lower Confidence) | Predicted "create" (7.46%), "help" (6.57%), "provide" (6.09%) with lower confidence scores | BART supports MLM but is primarily trained for seq2seq tasks; less optimized for fill-mask compared to BERT/RoBERTa |
| **QA** | BERT | Success (Poor Quality) | Extracted "and deepfakes." with very low confidence (0.50%) | BERT-base was not fine-tuned on SQuAD; the QA head is randomly initialized, leading to poor extractive performance |
| | RoBERTa | Success (Better) | Extracted "significant risks such as hallucinations, bias, and deepfakes" with low confidence (1.19%) | RoBERTa-base also not fine-tuned for QA; but its stronger encoder provides slightly better span extraction than BERT |
| | BART | Success (Best) | Extracted "Generative AI poses significant risks such as hallucinations," with highest confidence (3.71%) | BART's Encoder-Decoder architecture is more suitable for extractive tasks; provides better context understanding despite not being fine-tuned |

### Key Insights:
- **Text Generation**: All models technically "succeeded" but produced poor quality output. Encoder-only models (BERT/RoBERTa) are fundamentally unsuited for generation. BART has the right architecture but needs task-specific training.
- **Fill-Mask**: BERT and RoBERTa excel at this task (their training objective). BART works but with lower confidence since it's optimized for different tasks.
- **Question Answering**: Without fine-tuning on SQuAD, all models performed poorly but demonstrate architectural differences. BART's encoder-decoder structure gave it a slight edge.